pegen/pegen.c

```diff
@@ -221,44 +220,88 @@ keyword_token(Parser *p, const char *val)
 }
 
 PyObject *
-run_parser(const char *filename, void *(start_rule_func)(Parser *), int mode)
+run_parser(struct tok_state* tok, void *(start_rule_func)(Parser *), int mode)
 {
-    FILE *fp = fopen(filename, "rb");
-    if (fp == NULL)
-        panic("Can't open file");
-
-    Parser *p = malloc(sizeof(Parser));
+    int fail = 0;
+    Parser *p = PyMem_Malloc(sizeof(Parser));
     if (p == NULL)
         panic("Out of memory for Parser");
 
-    p->tok = PyTokenizer_FromFile(fp, NULL, NULL, NULL);
-    if (p->tok == NULL)
-        return NULL;
-
-    p->tokens = malloc(sizeof(Token));
+    assert(tok != NULL);
+    p->tok = tok;
+    p->tokens = PyMem_Malloc(sizeof(Token));
     memset(p->tokens, '\0', sizeof(Token));
     p->mark = 0;
     p->fill = 0;
     p->size = 1;
 
     p->arena = PyArena_New();
-    if (!p->arena)
-        return NULL;
+    if (!p->arena) {
+        fail = 1;
+        goto exit;
+    }
 
     fill_token(p);
 
     void *res = (*start_rule_func)(p);
     if (res == NULL) {
         PyErr_Format(PyExc_SyntaxError, "error at mark %d, fill %d, size %d", p->mark, p->fill, p->size);
+        fail = 1;
+        goto exit;
+    }
+
+exit:
+
+    PyMem_Free(p->tokens);
+
+    if (fail) {
+        PyArena_Free(p->arena);
```

Comment:
There is a bit of redundancy here on the cleanup for the arena and p, but I find this more readable than other versions that I tried factoring that out. Tell me if you have a preference in mind or some other structure that I have not consider.

Reply:
It's always "arena free; mem free; return" -- why not have a shared `result` variable and a `goto`?

Reply:
Indeed, I will take this as a sign that is too late for today :sweat_smile: